Feature/sim 2112/write end2end scripts

danielsf opened the pull request:
This is one of many (one for each sims_* repo...) PRs related to my effort to create a seamless integration between OpSim, CatSim, and PhoSim for generating full-focal plane image simulations.

This (sims_catalogs) pull request:

1) Changes how InstanceCatalog filters rows on `cannot_be_null` to be a little more efficient.

2) Adds a `parallel_catalog_writer()` which can write many InstanceCatalogs based on the same database connection (i.e. if you wanted to produce a PhoSim InstanceCatalog and a reference catalog of true magnitudes for the same objects without having to open the database connection twice).  This is analogous to the `CompoundInstanceCatalog`, except that the `CompoundInstanceCatalog` writes everything to one file.  The `parallel_catalog_writer` writes separate files.

3) Makes a cache of database connections that is a class attribute of `CatalogDBObject` which `CatalogDBObject` automatically uses.  Users no longer have to be careful about passing connections around to prevent redundant connections from opening (note that `DBObject` does not use the cache of connections).

danielsf committed:
run through cannot_be_null columns before all other columns

danielsf committed:
CatalogDBObjects keep a cache of open connections to draw from

danielsf committed:
use is rather than == when picking from connection_cache

danielsf committed:
make _get_connection importable from outside

danielsf committed:
CompoundCatalogDBObject instantiated sub-classes to make columns

this causes the columns that are a part of the database but not
explicitly declared in self.columns by the user to be ultimately
included in self.columns.  This should not cause extraneous db
connections to be open, so long as the global _connection_cache
in dbConnection.py is used by DBObject to open DBConnections

danielsf committed:
add unit test to check that we can filter catalogs with cannot_be_null

danielsf committed:
added test case for simple filter on InstanceCatalog

danielsf committed:
test cleans up after itself

danielsf committed:
add test for catalog filtered on two columns

danielsf committed:
default cannot_be_null is None, rather than an empty list

danielsf committed:
add test on filters added after instantiation

danielsf committed:
handle compound columns with new cannot_be_null scheme

danielsf committed:
add test for cannot_be_null with a compound column

danielsf committed:
cache getters in test so that we actually update the cache

danielsf committed:
test filtering on a compound column

danielsf committed:
unit test cleans up after itself

danielsf committed:
test a CompoundInstanceCatalog filtered on a compound column

danielsf committed:
make sure 0-length chunks are treated correctly

danielsf committed:
unit test cleans up after itself

danielsf committed:
if cannot compute filter column, raise exception

danielsf committed:
make it possible to filter on columns not written to the catalog

danielsf committed:
programmatic unittests all verify filter conditions explicitly

danielsf committed:
optional cannot_be_null kwarg to __init__()

danielsf committed:
add method to close all open database connections

danielsf committed:
unit test cleans up after itself

danielsf committed:
test what happens when you add kwarg cannot_be_null
to definition cannot_be_null

danielsf committed:
only CatalogDBObjects use the cache of database connections

danielsf committed:
add unit test for database connection cache

danielsf committed:
testConnectionPassing.py tests that passed connection is identical

danielsf committed:
add parallelCatalogWriter (not tested yet)

danielsf committed:
factor out chunk filtration into its own method

danielsf committed:
factor out the step where we update _current_chunk and _column_cache

danielsf committed:
forgot to update final_dexes in _filter_chunk()

danielsf committed:
add unit test for parallel catalog writer

danielsf committed:
add boolean to turn off header to parallelCatalogWriter

danielsf committed:
parallelCatalogWriter takes instantiations, rather than classes

danielsf committed:
make sure that all of the catalogs have the same obs_metadata

danielsf committed:
only pass constraint and chunk_size through parallelCatalogWriter

danielsf committed:
also test db_obj on objid in ParallelCatalogWriter

danielsf committed:
include objid in the error message when parallelCatalogWriter fails

danielsf committed:
fix typo

danielsf committed:
amend error message

danielsf committed:
timing print statements (for diagnostics)

danielsf committed:
parallelCatalogWriter applies cannot_be_null to catalogs individually

i.e. not every catalog will have the same rows, if their
cannot_be_nulls are disjoint

danielsf committed:
move connection_cache into CatalogDBObject

danielsf committed:
remove print statement

danielsf committed:
fix bug in which ParalleCatalogWriter was always appending to catalogs

danielsf committed:
testParallelCatalogWriter.py tests chunk_size != None

danielsf committed:
make CatalogDBObject incompleteness message more informative

danielsf committed:
cast all attributes as str when testing if we can use connection cache

danielsf committed:
cast all attributes as str in DBConnection.__eq__

danielsf committed:
remove print statement

jbkalmbach commented on python/lsst/sims/catalogs/definitions/ParallelCatalogWriter.py:
```diff
@@ -0,0 +1,113 @@
+    catalog_dict is a dict keyed on the names of the files to be written.
+    The values are the IntanceCatalogs to be written (note: these are full
+    instantiations of InstanceCatalogs, not just InstanceCatalog classes
+    as with the CompoundInstanceCatalog)
```
Looking at the code I don't think there's anything that prevents including CompoundInstanceCatalogs in here too, right? If so, maybe add that it can also include an instantiation of a compoundInstanceCatalog.

danielsf replied:
It won't work.  The `parallelCatalogWriter` loops over all of the catalogs calling `catalog._write_current_chunk()` for each chunk in the database query.  `CompoundInstanceCatalog` does not have a `_write_current_chunk()` method (among other things).  We could probably re-factor `CompoundInstanceCatalog` to look like something the `parallelCatalogWriter` could handle in the future, if we decide that would be useful.

Are you saying you want access to this for Twinkles work?

jbkalmbach commented on python/lsst/sims/catalogs/definitions/ParallelCatalogWriter.py:
```diff
@@ -0,0 +1,113 @@
+    Parameters
+    ----------
+    catalog_dict is a dict keyed on the names of the files to be written.
+    The values are the IntanceCatalogs to be written (note: these are full
```
typo InstanceCatalogs

jbkalmbach commented on python/lsst/sims/catalogs/definitions/ParallelCatalogWriter.py:
```diff
@@ -0,0 +1,113 @@
+                raise RuntimeError('Catalogs passed to parallelCatalogWriter have different '
+                                   'ObservationMetaData.  I do not know how to deal with that.')
+
+            try:
```
Why don't you use the __eq__ method you created for DBConnection here?

danielsf replied:
Mostly because `DBConnection.__eq__` also tests on `verbose`.  I removed `verbose` from `DBConnection.__eq__` and used `==` here.

jbkalmbach commented on tests/testFilteringCatalogs.py:
```diff
@@ -0,0 +1,591 @@
+            os.unlink(cat_name)
+
+    def test_filter_on_compound_column(self):
+        """
```
What about a test on a compound catalog where (using the 'a','b','c' you have here) the column 'c' has null values, but 'a' or 'b' cannot be null and testing that nothing gets filtered? I admit I'm not sure what all gets used as a compound column in practice, but I see this as a possible theoretical case that hasn't been tested.

danielsf replied:
Good test, but that's not how `cannot_be_null` works, now.  `cannot_be_null` will filter on a column even if that column is not actually written to the catalog.  That way, you can filter a catalog on the r magnitude without having to actually write the r magnitude to your catalog.

I will clarify the docstring.

jbkalmbach commented on tests/testParallelCatalogWriter.py:
```diff
@@ -0,0 +1,246 @@
+    host = None
+    port = None
+    driver = 'sqlite'
+    objid = 'paralle_writer_test_db'
```
Stupid comment, but did you mean this to be parallel_writer_test_db?

jbkalmbach commented on tests/testParallelCatalogWriter.py:
```diff
@@ -0,0 +1,246 @@
+    def get_valid3(self):
+        ii = self.column_by_name('id')
+        return np.where(ii%5 == 0, ii, None)
+
```
This goes with my earlier comment, but do we not want to be able to use this with CompoundInstanceCatalogs?

danielsf replied:
I will look into what is involved with making the `parallelCatalogWriter` work with a `CompoundInstanceCatalog`.

danielsf replied:
This should probably be work for a future ticket.  It is not trivial since `CompoundInstanceCatalog` allows you to pass in `InstanceCatalog`s that use potentially different database connections and then does the work of grouping them by like connection.  `parallelCatalogWriter` assumes that everything you pass it relies on the same database connection.

jbkalmbach replied:
Yeah, don't worry about it now, I didn't recall how much different `CompoundInstanceCatalog` and `InstanceCatalog` were.

jbkalmbach commented on python/lsst/sims/catalogs/definitions/InstanceCatalog.py:
```diff
@@ -254,6 +235,18 @@ def __init__(self, db_obj, obs_metadata=None, column_outputs=None,
                     if col not in self._column_outputs:
                         self._column_outputs.append(col)
 
+        self._cannot_be_null = None
```
The three different forms of "cannot_be_null" makes this hard to understand at times if you lose track of which is which. Maybe adding a couple comments about what's going on here can help keep them straight.

jbkalmbach reviewed (approved):
Looks good, just a few small suggestions.

danielsf committed:
fix typo

danielsf committed:
clarify docstring

danielsf committed:
parallelCatalogWriter tests connections with ==

part of this change is making DBConnection.__eq__ ignore the
verbose attribute.  This is probably harmless.  Verbosity does
not affect what the DBConnection is connecting to.

danielsf committed:
test column filtering on an unused column

danielsf committed:
clarify docstring

danielsf committed:
fix spelling error

danielsf committed:
add comment explaining why we had to add self._cannot_be_null
